### ISSUE TITLE
IndaBOM Installation

### ISSUE THREAD

**Issue by louisyssel** [mpkasp/indabom#3]
Hi there,

Firstly, I am NOT a developer, and therefore not to clever when it comes to django. I am however an infrastructre person, and do understand operating systems etc.

I ran IndaBOM on your server to validate the functionality, and it is really what I wanted all along, however, I simply cannot get it to run, as the instructions are a bot confusing.

I have a complete clean machine, with Ubuntu 18.04.

What are the step by step instructions to install IndaBOM on this machine please!!!!

What are the pre-requisites? (mysql, java, .......)

Then how to get Instabom running to be the same as on your cloud site.

I would really appreciate the help, as follwing the instruction you provided, sdoes NOT result in a running application.

Thanks
Louis

**Comment by louisyssel** [comment]
Bump

**Comment by mpkasp** [comment]
Hi Louis,

Can you provide more information? E.g.

What step are you failing on?
What error messages are you seeing?
What have you tried?

Mike

**Comment by louisyssel** [comment]
Hi Mike, 

Thanks for getting back to me.

1. I tried installing following your details on Centos 7, and after trying multiple times trying to add things that I could figure out from the error messages, but then still fails on the python manage.py runserver

2. I then installed a clean machine with Ubuntu 18.04 bionix. Folloed the steps again, and eventually got it to start, but when going into the web interface, I get errors on the web interface that it cannot find bom/ or admin/.
I removed the admin from the settings.py file, but then still fails.

3. I then reinstalled the machine again, and tried again, as I though maybe something could have screwed up the installation, and tried from scratch. This time all went well, until after "pip install -r requirements.txt, which finished normally. I then ran python manage.py migrate, and now I get the following error.
(bom) root@netmon:/opt/django/apps/django/bom/bom# python manage.py migrate
Traceback (most recent call last):
  File "manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/base.py", line 316, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/base.py", line 350, in execute
    self.check()
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/base.py", line 379, in check
    include_deployment_checks=include_deployment_checks,
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 60, in _run_checks
    issues.extend(super()._run_checks(**kwargs))
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/management/base.py", line 366, in _run_checks
    return checks.run_checks(**kwargs)
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/checks/registry.py", line 71, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/checks/urls.py", line 40, in check_url_namespaces_unique
    all_namespaces = _load_all_namespaces(resolver)
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/checks/urls.py", line 67, in _load_all_namespaces
    namespaces.extend(_load_all_namespaces(pattern, current))
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/checks/urls.py", line 67, in _load_all_namespaces
    namespaces.extend(_load_all_namespaces(pattern, current))
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/checks/urls.py", line 67, in _load_all_namespaces
    namespaces.extend(_load_all_namespaces(pattern, current))
  [Previous line repeated 980 more times]
  File "/opt/django/apps/django/bom/lib/python3.6/site-packages/django/core/checks/urls.py", line 59, in _load_all_namespaces
    ':'.join(parents + (url.namespace,)) for url in url_patterns
RecursionError: maximum recursion depth exceeded

This recursion error was also an error I experienced when I got the web interface up previously.

I really need to get this up and running, and this is the exact functionality we need, and clearly I am missing something.

Thanks
Louis

**Comment by mpkasp** [comment]
Hey Louis,

Thanks for the added details. Have you checked out the django-bom repository? I may need to add a note to the readme, but this repository is a web app that uses the core BOM tool. The actual BOM tool is implemented here: https://github.com/mpkasp/django-bom, and has a  lot more detailed instructions to get up and running. 

Also one of our members created a great video of how to get up and running. It may be a bit outdated now, but should have some good pointers:  https://indabom.com/install/

Let me know if that helps.

Mike

**Comment by louisyssel** [comment]
Thanks Mike, I will look into it tonight, and let you know.

**Comment by louisyssel** [comment]
Hi Mike,

I have now tried exactly what you recommended, and I followed the instructions on the video of the link you supplied.

1. I installed VirtualBox as per the video. (Previously I was running on a dedicated server.)
2. I installed Ubuntu 16.08 (Previously I was running Ubuntu 18.04)
3. I followed his steps exactly to the tee.
4. Everything was the same until where I had to run python manage.py migrate, or as he suggested, skipping it, and installing django-bom, and then again tried "python manage.py migrate' as well as "python manage.py createsuperuser.

This is the error I get here:
(zbom) osboxes@osboxes:~/zbom/zbom$ python manage.py migrate
^[^[^[^[Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/base.py", line 327, in execute
    self.check()
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/base.py", line 359, in check
    include_deployment_checks=include_deployment_checks,
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 62, in _run_checks
    issues.extend(super(Command, self)._run_checks(**kwargs))
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/management/base.py", line 346, in _run_checks
    return checks.run_checks(**kwargs)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/checks/registry.py", line 81, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/checks/urls.py", line 16, in check_url_config
    return check_resolver(resolver)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/core/checks/urls.py", line 26, in check_resolver
    return check_method()
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/urls/resolvers.py", line 254, in check
    for pattern in self.url_patterns:
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/urls/resolvers.py", line 405, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/urls/resolvers.py", line 398, in urlconf_module
    return import_module(self.urlconf_name)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/osboxes/zbom/zbom/zbom/urls.py", line 22, in <module>
    url(r'^bom/', include('bom.urls')),
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/django/conf/urls/__init__.py", line 50, in include
    urlconf_module = import_module(urlconf_module)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/osboxes/zbom/local/lib/python2.7/site-packages/bom/urls.py", line 2, in <module>
    from django.urls import path
ImportError: cannot import name path
I really do not know what to try next. Your help will really be appreciated.

Louis

**Comment by mpkasp** [comment]
Great! Progress.

1) Make sure you're running Python 3.7.0
if you're using a virtualenvironment, make sure you do the step in the instructions that specifies using python 3 `virtualenv -p python3 mysite`

2) Make sure you're running the correct version of django as specified in the requirements.txt file: Django==2.1.5

**Comment by louisyssel** [comment]
OK, looks better already, but I do not have a GUI on the server to test now, how can I change the runserver to listen on a different address than 127.0.0.1 and port?

**Comment by mpkasp** [comment]
Great!!

You can specify the ip address and port in the manage.py command I think, maybe something like this?

`python ./manage.py runserver <your_ip_address>:<your_port>`

If you want to set it up to be accessible beyond your local machine, I'd recommend looking into how to set up a django project on whatever cloud service provider you want to use (e.g. AWS, Azure, or Google Cloud).

**Comment by louisyssel** [comment]
No worries, I figured our changing the IP and port, I can access Admin perfectly, and perform all of the Admin functions.

When I want to go to bom hhowever, this is what I get.
<img width="776" alt="image" src="https://user-images.githubusercontent.com/29566427/53206481-bdfcae00-3639-11e9-91e6-254ad0be0553.png">

**Comment by mpkasp** [comment]
Try heading to the admin and check your organizations, you may have more than one tied to your user, which doesn't work when the get() function is called.

**Comment by louisyssel** [comment]
Excellent, I am up and running. Thank you so much for your help.

PS. Do you have a manual for the system so I can understand the relationships between Manuf parts, and Part Classes and Part files etc?

Thanks again

**Comment by louisyssel** [comment]
Oh, last problem:

When I am in the admin site, and I click on the right corner on "VIEW SITE" I get the error below, itseems to not add the /bom/ at the end:

<img width="914" alt="image" src="https://user-images.githubusercontent.com/29566427/53207362-2f3d6080-363c-11e9-8197-81e5e1f0f871.png">

**Comment by louisyssel** [comment]
Apologies for the million questions??? Where can I change the currency from USD to ZAR?

**Comment by mpkasp** [comment]
No problem. I don't have a manual - the project started as an internal tool and I figured I'd release it to the world as I figured others would be interested! I haven't had time to document anything yet, and it's also the tool is continuing to evolve.

Yeah I never fixed that "view site" link. If you set up django-bom on the indabom project and use the indabom site admin, it should handle redirect a root / path.

Finally on your last question one of the roadmap features is to support different currencies, however I haven't pulled that in yet. I created an issue on the django-bom page so it's tracked: https://github.com/mpkasp/django-bom/issues/10
